File: src/reach/review.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Provide ephemeral HTTP server and browser UI for interactive query boundary curation."""

from __future__ import annotations

import contextlib
import functools
import html
import http.server
import json
import os
import select
import socketserver
import sys
import threading
import time
import urllib.parse
import webbrowser
from dataclasses import dataclass, field
from enum import StrEnum
from typing import TYPE_CHECKING, Annotated, ClassVar, Self, cast, override

from reach.models import Query, QueryKind, Skill
from reach.queries import QuerySet, QuerySetProvenance
from reach.static import get_review_css, get_review_js, get_review_template
from reach.view import _esc
from reach.views import build_console

if TYPE_CHECKING:
    import socket
    from collections.abc import Sequence

from pydantic import (
    BaseModel,
    ConfigDict,
    StringConstraints,
    ValidationInfo,
    field_validator,
    model_validator,
)

__all__ = [
    "ReviewSentinel",
    "ReviewServerHandler",
    "launch_query_review",
    "render_query_review_html",
]


class ReviewSentinel(StrEnum):
    """Sentinel values used across the interactive review interface."""

    OUT_OF_SCOPE = "__OUT_OF_SCOPE__"


#: Inlined CSS styling for the split territory boundary curation interface.
_REVIEW_STYLE = get_review_css()

#: Inlined client-side JavaScript for keyboard navigation and interactive curation.
_REVIEW_SCRIPT = get_review_js()

#: Allowed hostnames for loopback / local review server validation.
_ALLOWED_HOSTS: frozenset[str] = frozenset({"127.0.0.1", "localhost", "testserver", "::1"})


def render_query_review_html(
    query_set: QuerySet,
    skill: Skill,
    rivals: Sequence[Skill],
) -> str:
    """Render an interactive standalone HTML boundary curation interface for a QuerySet."""
    rival_names = [r.name for r in rivals if r.name != skill.name]
    target_name = skill.name

    trigger_cards = []
    guardrail_cards = []

    for q in query_set.queries:
        exp = q.expected_skill
        is_trigger = exp == target_name
        kind_name = q.kind.value if q.kind else "implicit"

        oos_val = ReviewSentinel.OUT_OF_SCOPE.value
        rival_options = [f'<option value="{oos_val}">Out of Scope (General Distractor)</option>']
        for r_name in rival_names:
            sel = ' selected="selected"' if exp == r_name else ""
            rival_options.append(
                f'<option value="{_esc(r_name)}"{sel}>Rival: {_esc(r_name)}</option>'
            )

        rival_select_disp = "none" if is_trigger else "block"
        swap_btn_text = "⇄ Move to Guardrails" if is_trigger else "⇄ Move to Triggers"

        if is_trigger:
            badge_target = f'<span class="badge-target">→ {_esc(target_name)}</span>'
        elif exp and exp in rival_names:
            badge_target = f'<span class="badge-rival">→ {_esc(exp)}</span>'
        else:
            badge_target = '<span class="badge-distractor">→ Out of Scope</span>'

        card_html = (
            f'<div class="query-card">'
            f'  <div class="card-top">'
            f'    <div class="card-badges">'
            f'      <span class="badge-kind">{_esc(kind_name)}</span>'
            f'      <span class="badge-target-wrapper">{badge_target}</span>'
            f"    </div>"
            f'    <button type="button" class="card-btn card-btn-delete"'
            f' onclick="deleteCard(this)" title="Delete query">✕</button>'
            f"  </div>"
            f'  <div class="query-preview query-text-preview"'
            f' onclick="editCard(this)">{_esc(q.text)}</div>'
            f'  <div class="query-card-editor">'
            f'    <textarea class="query-input"'
            f' placeholder="Enter query prompt...">{_esc(q.text)}</textarea>'
            f'    <select class="rival-select" style="display:{rival_select_disp};"'
            f' onchange="updateCardRival(this)">'
            f"      {''.join(rival_options)}"
            f"    </select>"
            f"  </div>"
            f'  <div class="card-actions">'
            f'    <div class="card-left-actions">'
            f'      <button type="button" class="card-btn card-btn-swap"'
            f' onclick="swapCard(this)">{swap_btn_text}</button>'
            f"    </div>"
            f'    <div class="card-right-actions">'
            f'      <button type="button" class="card-btn card-btn-edit"'
            f' onclick="editCard(this)">✎ Edit</button>'
            f'      <button type="button" class="card-btn card-btn-done"'
            f' onclick="closeCard(this)">✓ Done</button>'
            f"    </div>"
            f"  </div>"
            f"</div>"
        )

        if is_trigger:
            trigger_cards.append(card_html)
        else:
            guardrail_cards.append(card_html)

    rivals_json = html.escape(json.dumps(rival_names), quote=True)
    initial_count = len(query_set.queries)
    empty_style = "display: none;" if initial_count > 0 else ""
    footer_text = (
        f"{initial_count} queries total: {len(trigger_cards)} should trigger, "
        f"{len(guardrail_cards)} should not trigger"
    )

    rival_chips = "".join(f'<span class="rival-chip">{_esc(r)}</span>' for r in rival_names)
    no_rivals_msg = (
        '<div class="rivals-bar"><span>No competing skills detected in neighborhood.</span></div>'
    )
    rivals_section = (
        f'<div class="rivals-bar"><span>Competing rivals:</span> {rival_chips}</div>'
        if rival_names
        else no_rivals_msg
    )

    trig_pct = (len(trigger_cards) / (initial_count or 1)) * 100
    guard_pct = (len(guardrail_cards) / (initial_count or 1)) * 100
    trig_empty_style = "display:none;" if trigger_cards else ""
    guard_empty_style = "display:none;" if guardrail_cards else ""

    return get_review_template().format(
        skill_name=_esc(skill.name),
        style=_REVIEW_STYLE,
        target_skill=_esc(target_name),
        rivals_json=rivals_json,
        out_of_scope=ReviewSentinel.OUT_OF_SCOPE.value,
        skill_desc=_esc(skill.description),
        rivals_section=rivals_section,
        summary=footer_text,
        trig_pct=f"{trig_pct:.1f}",
        guard_pct=f"{guard_pct:.1f}",
        triggers_count=len(trigger_cards),
        triggers_cards="".join(trigger_cards),
        trig_empty_style=trig_empty_style,
        guardrails_count=len(guardrail_cards),
        guardrails_cards="".join(guardrail_cards),
        guard_empty_style=guard_empty_style,
        empty_style=empty_style,
        script=_REVIEW_SCRIPT,
    )


class _ReviewQueryItem(BaseModel):
    """Represent a single query item received from the curation UI."""

    # UI payload flexibility: UI may submit extraneous metadata or layout attributes
    model_config = ConfigDict(frozen=True, extra="ignore")

    text: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
    expected_skill: str | None = None

    @field_validator("expected_skill", mode="before")
    @classmethod
    def _empty_is_absent(cls, value: object) -> object:
        """Convert empty strings to None for optional skill names."""
        if isinstance(value, str) and not value.strip():
            return None
        return value

    @model_validator(mode="after")
    def _validate_expected_skill(self, info: ValidationInfo) -> Self:
        """Validate expected skill name against catalog scope when context is available."""
        if info.context and "allowed_skills" in info.context:
            allowed = info.context["allowed_skills"]
            if self.expected_skill is not None and self.expected_skill not in allowed:
                msg = f"expected_skill {self.expected_skill!r} is not in allowed catalog skills"
                raise ValueError(msg)
        return self


class _ReviewPayload(BaseModel):
    """Represent the JSON payload submitted by the curation UI upon approval."""

    # UI payload flexibility: browser JSON may include extra client-side state
    model_config = ConfigDict(frozen=True, extra="ignore")

    queries: tuple[_ReviewQueryItem, ...]


@dataclass
class _ReviewSession:
    """Encapsulate session-specific state for an ephemeral review server."""

    html_content: str
    done_event: threading.Event = field(default_factory=threading.Event)
    saved_queries: tuple[_ReviewQueryItem, ...] | None = None
    allowed_skills: set[str] = field(default_factory=set)


class ReviewServerHandler(http.server.BaseHTTPRequestHandler):
    """Handle HTTP requests for the ephemeral interactive review server."""

    html_content: ClassVar[str] = ""
    saved_queries: ClassVar[tuple[_ReviewQueryItem, ...] | None] = None
    done_event: ClassVar[threading.Event | None] = None
    allowed_skills: ClassVar[set[str]] = set()

    def __init__(
        self,
        request: socket.socket | tuple[bytes, socket.socket],
        client_address: tuple[str, int] | str,
        server: socketserver.BaseServer,
        *,
        session: _ReviewSession | None = None,
    ) -> None:
        """Initialize the request handler with an optional bound review session."""
        self.session = session
        super().__init__(request, client_address, server)

    def _is_valid_host(self) -> bool:
        """Validate Host header to prevent DNS rebinding attacks."""
        host = self.headers.get("Host", "")
        if not host:
            return True
        try:
            parsed = urllib.parse.urlsplit(f"//{host}")
            hostname = (parsed.hostname or "").lower()
        except ValueError:
            return False
        return hostname in _ALLOWED_HOSTS

    def _is_valid_origin(self) -> bool:
        """Validate Origin header on mutating requests to protect against CSRF."""
        origin = self.headers.get("Origin")
        if not origin:
            return True
        if origin == "null":
            return True
        try:
            parsed = urllib.parse.urlsplit(origin)
        except ValueError:
            return False
        if parsed.scheme not in ("http", "https"):
            return False
        hostname = (parsed.hostname or "").lower()
        return hostname in _ALLOWED_HOSTS

    def _send_security_headers(self) -> None:
        """Add defensive security headers to all responses."""
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("X-Frame-Options", "DENY")

    def _send_json(self, status: int, payload: object) -> None:
        """Serialize and transmit a JSON response with standard security headers."""
        resp = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(resp)))
        self._send_security_headers()
        self.end_headers()
        self.wfile.write(resp)

    def do_GET(self) -> None:
        """Serve the review HTML page or health status."""
        if not self._is_valid_host():
            self.send_error(403, "Forbidden: Invalid Host header")
            return

        if self.path in ("/", "/index.html"):
            html_source = self.session.html_content if self.session else self.html_content
            encoded = html_source.encode("utf-8")
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("Content-Length", str(len(encoded)))
            self._send_security_headers()
            self.end_headers()
            self.wfile.write(encoded)
        elif self.path == "/api/status":
            self._send_json(200, {"status": "running"})
        else:
            self.send_error(404, "Not Found")

    def do_POST(self) -> None:
        """Receive and validate curated queries from the browser."""
        if not self._is_valid_host():
            self.send_error(403, "Forbidden: Invalid Host header")
            return

        if not self._is_valid_origin():
            self.send_error(403, "Forbidden: Cross-origin request rejected")
            return

        if self.path == "/api/save":
            try:
                content_len = int(self.headers.get("Content-Length", 0))
                if content_len < 0 or content_len > 5 * 1024 * 1024:
                    msg = "Content-Length must be between 0 and 5MB"
                    raise ValueError(msg)
            except (ValueError, TypeError) as exc:
                self._send_json(400, {"status": "error", "message": str(exc)})
                return

            post_body = self.rfile.read(content_len)
            try:
                data = json.loads(post_body.decode("utf-8"))
                allowed = self.session.allowed_skills if self.session else self.allowed_skills
                context = {"allowed_skills": allowed} if allowed else None
                payload = _ReviewPayload.model_validate(data, context=context)
            except (json.JSONDecodeError, ValueError) as exc:
                self._send_json(400, {"status": "error", "message": str(exc)})
                return

            # Save queries and unblock review loop
            if self.session is not None:
                self.session.saved_queries = payload.queries
                self.session.done_event.set()
            else:
                ReviewServerHandler.saved_queries = payload.queries
                if ReviewServerHandler.done_event:
                    ReviewServerHandler.done_event.set()

            self._send_json(200, {"status": "ok", "saved": len(payload.queries)})
        else:
            self.send_error(404, "Not Found")

    @override
    def log_message(self, format: str, *args: object) -> None:
        """Suppress standard HTTP request logging to keep console clean."""


def _convert_saved_queries(
    saved: Sequence[_ReviewQueryItem],
    target_skill_name: str,
) -> list[Query]:
    """Convert curated query items into validated Query instances."""
    updated_queries: list[Query] = []
    for i, item in enumerate(saved):
        text = item.text
        expected_skill = item.expected_skill

        if expected_skill == target_skill_name:
            kind = QueryKind.IMPLICIT
        elif expected_skill:
            kind = QueryKind.NEIGHBOR_NEGATIVE
        else:
            kind = QueryKind.OUT_OF_SCOPE

        updated_queries.append(
            Query(
                id=f"curated-{i + 1}",
                text=text,
                expected_skill=expected_skill,
                kind=kind,
            )
        )
    return updated_queries


def launch_query_review(
    query_set: QuerySet,
    skill: Skill,
    rivals: Sequence[Skill],
    *,
    timeout: float = 600.0,
) -> QuerySet:
    """Launch interactive browser review on an ephemeral HTTP server on 127.0.0.1.

    If in a headless or non-interactive environment, the review is gracefully bypassed.
    """
    console = build_console()

    # Bypass review if non-interactive, explicitly in CI, or REACH_NO_BROWSER is set
    no_browser = os.environ.get("REACH_NO_BROWSER", "").lower() in ("true", "1")
    is_ci = os.environ.get("CI", "").lower() in ("true", "1")
    if not sys.stdin.isatty() or is_ci or no_browser:
        n_queries = len(query_set.queries)
        console.print(
            f"[dim]Non-interactive environment; proceeding with {n_queries} queries.[/dim]"
        )
        return query_set

    # Render HTML content with target skill and rival options
    html_content = render_query_review_html(query_set, skill, rivals)

    allowed_skills = {skill.name, *(r.name for r in rivals)}
    session = _ReviewSession(
        html_content=html_content,
        allowed_skills=allowed_skills,
    )

    handler = functools.partial(ReviewServerHandler, session=session)
    server = http.server.ThreadingHTTPServer(
        ("127.0.0.1", 0),
        cast(type[http.server.BaseHTTPRequestHandler], handler),
    )
    server_port = server.server_port
    url = f"http://127.0.0.1:{server_port}"

    server_thread = threading.Thread(target=server.serve_forever, daemon=True)
    server_thread.start()

    console.print(
        f"[bold blue]Launched interactive boundary review on[/bold blue] [link={url}]{url}[/link]"
    )
    console.print(
        "[dim]Press [Enter] in this terminal to proceed immediately with current queries.[/dim]"
    )

    with contextlib.suppress(Exception):
        webbrowser.open(url)

    # Wait for either browser approval, terminal [Enter], or timeout
    start_time = time.monotonic()
    try:
        while not session.done_event.is_set():
            if time.monotonic() - start_time > timeout:
                console.print(
                    f"\n[dim]Review timed out after {int(timeout)}s; "
                    "proceeding with current queries.[/dim]"
                )
                break

            # Poll terminal stdin for Enter key with small timeout
            with contextlib.suppress(Exception):
                rlist, _, _ = select.select([sys.stdin], [], [], 0.2)
                if rlist:
                    line = sys.stdin.readline()
                    if line:
                        console.print("\n[dim]Proceeding from terminal approval...[/dim]")
                        break

            time.sleep(0.1)
    finally:
        server.shutdown()
        server_thread.join(timeout=2.0)
        server.server_close()

    # If queries were saved through the browser, convert them into QuerySet
    saved = session.saved_queries
    if saved is not None:
        updated_queries = _convert_saved_queries(saved, skill.name)
        if updated_queries:
            console.print(
                f"[bold green]✓[/bold green] Curated {len(updated_queries)} "
                f"queries for '{skill.name}'."
            )
            return QuerySet(
                catalog_id=query_set.catalog_id,
                provenance=QuerySetProvenance(origin=query_set.provenance.origin),
                queries=tuple(updated_queries),
            )

    return query_set

File: src/reach/static/review.js
/* Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

let activeCardIndex = -1;

function getTargetSkill() {
  return document.body.dataset.targetSkill || '';
}

function getOutOfScope() {
  return document.body.dataset.outOfScope || '__OUT_OF_SCOPE__';
}

function getAllCards() {
  return Array.from(document.querySelectorAll('.query-card'));
}

function setElementText(id, text) {
  const el = document.getElementById(id);
  if (el) el.textContent = text;
}

function setElementDisplay(id, display) {
  const el = document.getElementById(id);
  if (el) el.style.display = display;
}

function escapeHtml(str) {
  if (typeof str !== 'string') return '';
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}

function renderBadge(kind, text) {
  return `<span class="badge-${escapeHtml(kind)}">→ ${escapeHtml(text)}</span>`;
}

function setStatus(text, isError = false) {
  const statusBar = document.getElementById('status-bar');
  if (!statusBar) return;
  statusBar.style.color = isError ? 'var(--reach-error-text)' : 'var(--reach-success-text)';
  statusBar.textContent = text;
}

function updateSummary() {
  const triggers = document.querySelectorAll('#triggers-list .query-card').length;
  const guardrails = document.querySelectorAll('#guardrails-list .query-card').length;
  const total = triggers + guardrails;

  // Header badges
  setElementText('triggers-count', `${triggers} Triggers`);
  setElementText('guardrails-count', `${guardrails} Guardrails`);

  // Balance bar fill
  const totalSafe = total || 1;
  const triggerPct = (triggers / totalSafe) * 100;
  const guardrailPct = (guardrails / totalSafe) * 100;

  const barTriggers = document.getElementById('balance-triggers');
  if (barTriggers) barTriggers.style.width = `${triggerPct}%`;
  const barGuardrails = document.getElementById('balance-guardrails');
  if (barGuardrails) barGuardrails.style.width = `${guardrailPct}%`;

  // Summary text matching standard contract
  setElementText(
    'summary',
    `${total} queries total: ${triggers} should trigger, ${guardrails} should not trigger`,
  );

  // Column empty states
  setElementDisplay('triggers-empty', triggers === 0 ? 'flex' : 'none');
  setElementDisplay('guardrails-empty', guardrails === 0 ? 'flex' : 'none');

  // Overall empty state
  setElementDisplay('empty-state', total === 0 ? '' : 'none');
}

function getRows() {
  const cards = getAllCards();
  const targetSkill = getTargetSkill();
  const queries = [];

  cards.forEach((card) => {
    const input = card.querySelector('.query-input');
    if (!input) return;
    const text = input.value.trim();
    if (!text) return;

    const isTrigger = card.closest('#triggers-list') !== null;
    let expected = null;
    if (isTrigger) {
      expected = targetSkill;
    } else {
      const select = card.querySelector('.rival-select');
      if (select?.value && select.value !== getOutOfScope()) {
        expected = select.value;
      } else {
        expected = null;
      }
    }

    queries.push({
      text: text,
      expected_skill: expected,
    });
  });

  return queries;
}

function swapCard(btn) {
  const card = btn.closest('.query-card');
  if (!card) return;

  const isTrigger = card.closest('#triggers-list') !== null;
  const targetList = document.getElementById(isTrigger ? 'guardrails-list' : 'triggers-list');

  // Update card UI for new column
  const badgeTarget = card.querySelector('.badge-target-wrapper');
  const rivalSelect = card.querySelector('.rival-select');
  const swapBtn = card.querySelector('.card-btn-swap');

  if (isTrigger) {
    // Moving to guardrails
    if (badgeTarget) badgeTarget.innerHTML = renderBadge('rival', 'Rival / Out-of-Scope');
    if (rivalSelect) rivalSelect.style.display = 'block';
    if (swapBtn) swapBtn.textContent = '⇄ Move to Triggers';
  } else {
    // Moving to triggers
    if (badgeTarget) badgeTarget.innerHTML = renderBadge('target', getTargetSkill());
    if (rivalSelect) rivalSelect.style.display = 'none';
    if (swapBtn) swapBtn.textContent = '⇄ Move to Guardrails';
  }

  if (targetList) targetList.appendChild(card);
  updateSummary();
}

function deleteCard(btn) {
  const card = btn.closest('.query-card');
  if (card) {
    card.remove();
    updateSummary();
  }
}

function editCard(el) {
  const card = el.closest('.query-card');
  if (!card) return;
  card.classList.add('editing');
  const input = card.querySelector('.query-input');
  if (input) {
    input.focus();
    input.setSelectionRange(input.value.length, input.value.length);
  }
}

function closeCard(btn) {
  const card = btn.closest('.query-card');
  if (!card) return;
  const input = card.querySelector('.query-input');
  const preview = card.querySelector('.query-text-preview');
  if (input && preview) {
    preview.textContent = input.value.trim() || '(Empty query)';
  }
  card.classList.remove('editing');
  updateSummary();
}

function addRow(text = '', isTrigger = true, expected = '') {
  const targetSkill = getTargetSkill();
  const rivals = JSON.parse(document.body.dataset.rivals || '[]');

  const outOfScope = getOutOfScope();
  let rivalOptions = `<option value="${escapeHtml(outOfScope)}">Out of Scope (General Distractor)</option>`;
  for (const r of rivals) {
    const safeR = escapeHtml(r);
    const sel = expected === r ? 'selected' : '';
    rivalOptions += `<option value="${safeR}" ${sel}>Rival: ${safeR}</option>`;
  }

  const targetList = document.getElementById(isTrigger ? 'triggers-list' : 'guardrails-list');
  const card = document.createElement('div');
  card.className = 'query-card editing';

  const badgeHtml = isTrigger
    ? renderBadge('target', targetSkill)
    : renderBadge('rival', expected || 'Rival / Out-of-Scope');

  const rivalSelectDisp = isTrigger ? 'none' : 'block';
  const swapBtnText = isTrigger ? '⇄ Move to Guardrails' : '⇄ Move to Triggers';
  const safeText = escapeHtml(text);

  card.innerHTML = `
    <div class="card-top">
      <div class="card-badges">
        <span class="badge-kind">authored</span>
        <span class="badge-target-wrapper">${badgeHtml}</span>
      </div>
      <button type="button" class="card-btn card-btn-delete"
        onclick="deleteCard(this)" title="Delete query">✕</button>
    </div>
    <div class="query-text-preview" onclick="editCard(this)">${safeText ? safeText : '(Empty query)'}</div>
    <div class="query-card-editor">
      <textarea class="query-input" placeholder="Enter query prompt...">${safeText}</textarea>
      <select class="rival-select" style="display:${rivalSelectDisp};"
        onchange="updateCardRival(this)">
        ${rivalOptions}
      </select>
    </div>
    <div class="card-actions">
      <div class="card-left-actions">
        <button type="button" class="card-btn card-btn-swap"
          onclick="swapCard(this)">${swapBtnText}</button>
      </div>
      <div class="card-right-actions">
        <button type="button" class="card-btn card-btn-edit"
          onclick="editCard(this)">✎ Edit</button>
        <button type="button" class="card-btn card-btn-done"
          onclick="closeCard(this)">✓ Done</button>
      </div>
    </div>
  `;

  if (targetList) targetList.appendChild(card);
  updateSummary();
  const input = card.querySelector('.query-input');
  if (input) input.focus();
}

function updateCardRival(select) {
  const card = select.closest('.query-card');
  if (!card) return;
  const badgeTarget = card.querySelector('.badge-target-wrapper');
  if (!badgeTarget) return;

  const isOutOfScope = select.value === getOutOfScope();
  badgeTarget.innerHTML = isOutOfScope
    ? renderBadge('distractor', 'Out of Scope')
    : renderBadge('rival', select.value);
}

function exportEvalSetJson() {
  const queries = getRows();
  const targetSkill = getTargetSkill();
  const data = queries.map((q) => ({
    query: q.text,
    should_trigger: q.expected_skill === targetSkill,
    expected_skill: q.expected_skill,
  }));
  const blob = new Blob([JSON.stringify(data, null, 2)], { type: 'application/json' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = `eval_set_${targetSkill}.json`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

async function approveQueries() {
  const queries = getRows();
  if (queries.length === 0) {
    alert('Please keep at least one query in the eval set.');
    return;
  }
  setStatus('Saving queries and resuming optimizer in terminal...');

  const token =
    document.querySelector('meta[name="reach-token"]')?.content ||
    new URLSearchParams(window.location.search).get('token') ||
    '';
  const headers = { 'Content-Type': 'application/json' };
  if (token) {
    headers['X-Reach-Token'] = token;
  }

  try {
    const resp = await fetch('/api/save', {
      method: 'POST',
      headers: headers,
      body: JSON.stringify({ queries: queries }),
    });
    if (resp.ok) {
      setStatus('✓ Saved! You can close this tab and return to the terminal.');
      document.querySelectorAll('.approve-btn').forEach((btn) => {
        btn.disabled = true;
      });
    } else {
      setStatus(`Error saving queries: ${resp.statusText}`, true);
    }
  } catch (err) {
    setStatus(`Network error saving queries: ${err}`, true);
  }
}

// Keyboard Navigation & Shortcuts
document.addEventListener('keydown', (e) => {
  if ((e.metaKey || e.ctrlKey) && e.key === 'Enter') {
    e.preventDefault();
    approveQueries();
    return;
  }

  // If typing in an active input or textarea, don't trigger hotkeys
  if (['TEXTAREA', 'INPUT', 'SELECT'].includes(document.activeElement.tagName)) {
    if (e.key === 'Escape') {
      const activeCard = document.activeElement.closest('.query-card');
      if (activeCard) closeCard(activeCard);
    }
    return;
  }

  const cards = getAllCards();
  if (cards.length === 0) return;

  if (e.key === 'j' || e.key === 'ArrowDown') {
    e.preventDefault();
    activeCardIndex = (activeCardIndex + 1) % cards.length;
    highlightActiveCard(cards);
    return;
  }
  if (e.key === 'k' || e.key === 'ArrowUp') {
    e.preventDefault();
    activeCardIndex = (activeCardIndex - 1 + cards.length) % cards.length;
    highlightActiveCard(cards);
    return;
  }

  const activeCard = cards[activeCardIndex];
  if (!activeCard) return;

  if (e.key === ' ' || e.key === 't') {
    e.preventDefault();
    swapCard(activeCard.querySelector('.card-btn-swap'));
  } else if (e.key === 'Enter') {
    e.preventDefault();
    editCard(activeCard);
  } else if (e.key === 'Backspace' || e.key === 'Delete') {
    e.preventDefault();
    deleteCard(activeCard.querySelector('.card-btn-delete'));
    activeCardIndex = Math.min(activeCardIndex, getAllCards().length - 1);
  }
});

function highlightActiveCard(cards) {
  cards.forEach((c, idx) => {
    if (idx === activeCardIndex) {
      c.classList.add('active-card');
      c.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
    } else {
      c.classList.remove('active-card');
    }
  });
}

// Click to focus active card
document.addEventListener('click', (e) => {
  const card = e.target.closest('.query-card');
  if (card) {
    const cards = getAllCards();
    activeCardIndex = cards.indexOf(card);
    highlightActiveCard(cards);
  }
});

// Initial calculation and card highlight
window.addEventListener('DOMContentLoaded', () => {
  updateSummary();
  const cards = getAllCards();
  if (cards.length > 0) highlightActiveCard(cards);
});

// Explicitly bind functions invoked via HTML event attributes to window
window.addRow = addRow;
window.updateCardRival = updateCardRival;
window.exportEvalSetJson = exportEvalSetJson;
window.approveQueries = approveQueries;
window.deleteCard = deleteCard;
window.editCard = editCard;
window.closeCard = closeCard;
window.swapCard = swapCard;
